pm/sendermonitor.go

```diff
@@ -19,28 +22,24 @@ var unixNow = func() int64 {
 // SenderMonitor is an interface that describes methods used to
 // monitor remote senders
 type SenderMonitor interface {
-	RedeemableEmitter
-
 	// Start initiates the helper goroutines for the monitor
 	Start()
 
 	// Stop signals the monitor to exit gracefully
 	Stop()
 
 	// QueueTicket adds a ticket to the queue for a remote sender
-	QueueTicket(addr ethcommon.Address, ticket *SignedTicket) error
-
-	// AddFloat adds to a remote sender's max float
-	AddFloat(addr ethcommon.Address, amount *big.Int) error
-
-	// SubFloat subtracts from a remote sender's max float
-	SubFloat(addr ethcommon.Address, amount *big.Int)
+	QueueTicket(ticket *SignedTicket) error
 
 	// MaxFloat returns a remote sender's max float
 	MaxFloat(addr ethcommon.Address) (*big.Int, error)
 
 	// ValidateSender checks whether a sender's unlock period ends the round after the next round
 	ValidateSender(addr ethcommon.Address) error
+
+	// MonitorMaxFloat listens for max floats updates for a sender
+	// This method only needs to be "implemented" when the SenderMonitor consumer is a fire-and-forget module for winning tickets
```

Comment:
Isn't all ticket queuing now fire-and-forget from the perspective of the SenderMonitor consumer? Do you mean when the SenderMonitor itself is a remote redeemer? (If so, isn't float monitoring still necessary for local redeeming?)

Reply:
There's a difference here between `senderMonitor` and `SenderMonitor` (I know this gets confusing !). 

The `SenderMonitor` is implemented by `senderMonitor` and `redeemerClient` and is consumed by `recipient` or `Redeemer` (server). 

While `recipient` would be fire and forget, the ticket redemption _server_  is not as it needs to know when the `senderMonitor` updates `maxFloat` for a sender, to send an update to the `redeemerClients` accordingly. Since `AddFloat` and `SubFloat` are no longer part of the `SenderMonitor` API , consumers of this package have no other way of knowing when maxFloat gets updated. 

From discord: 
![image](https://user-images.githubusercontent.com/22256858/83077511-d9920800-a077-11ea-9553-c947a8b4110c.png)

Reply:
Hmm, this might just be an implementation detail that's going over my head, but is my understanding correct of the three possible "modes" here?

1. Orchestrator doing local redemptions, using DB backed queue
1. Orchestrator doing RPC redemptions, presumably this is what's fire-and-forget
1. RPC redemption server, using DB backed queue (hits the same code path as the O doing local redemptions, after receiving a ticket from RPC)

Don't all of these need to monitor for the max float? (This is where I'm not quite understanding why `MonitorMaxFloat` is only needed for the second mode, fire-and-forget orchestrators doing RPC redemptions [1].)

1. Orchestrator doing local redemptions still needs to monitor the blockchain if the B's float changes elsewhere, eg winning ticket on another node, or withdrawal
1. Receives float updates via RPC monitoring (this interface method?)
1. Receives float updates via blockchain *and* redemptions from other orchestrators. Notifies other orchestrators when this happens.

I guess the first and third modes "monitor" the float more implicitly? Spitballing here, but would it help to make this more explicit / symmetrical?

[1] My interpretation of "fire and forget" is that'd be a property of whatever's populating the ticket queue, whether via RPC call or DB insert. That transfers responsibility to whatever's responsible for pulling from the queue for submission, retries, etc. Am guessing that the "fire and forget" term has slightly different connotations here, referring only to the RPC submission method. But maybe that helps visualize how I see the symmetry between RPC and local submission.

Reply:
Another point of confusion here is the fact this takes a broadcaster address while the RPC float monitoring doesn't...

Reply:
>  is my understanding correct of the three possible "modes" here?

Yep that describes it well 

> Don't all of these need to monitor for the max float? 

No it depends on the consumer package and whether it needs outside knowledge about when a max float value is updated or not. Not to be confused with receiving updates over the wire protocol. 

in the first setup the `recipient` would be the consumer but that package doesn't care about tracking `maxFloat` it just needs to be able to access that value on the `SenderMonitor` through its API.  

In the third setup we already receive updated max float values through the wire protocol and also have access to a remote call if needed. 

The second setup however, the server, needs to know when the `SenderMonitor` abstraction does internal state updates for its max float so it can send updates over the wire to the clients.

I believe @yondonfu posted a suggestion to switch around the interfaces a bit to make this less confusing but I'm struggling to find it at the moment. 

> Another point of confusion here is the fact this takes a broadcaster address while the RPC float monitoring doesn't...

 I renamed this method to `SubscribeMaxFloat` if it helps

Reply:
> I believe @yondonfu posted a suggestion to switch around the interfaces a bit to make this less confusing but I'm struggling to find it at the moment.

Sorry I posted a comment, but then deleted it because I wanted to make some edits first. Re-posting below:

IIUC, `MonitorMaxFloat()` is only used by the `Redeemer` in #1513, but it is not required by the current consumers of the `SenderMonitor` API (i.e. `recipient` in the `pm` package). So, `MonitorMaxFloat()` doesn't actually need to be a part of the `SenderMonitor` API. However, we want to use `senderMonitor` in `Redeemer` since it contains the required ticket queuing and max float monitoring functionality. At the moment, we can't directly use `senderMonitor` in `Redeemer` since it is not exported and we only have access to the exported `SenderMonitor` interface. 

What if we did the following:

- Export `senderMonitor` and rename it to something like `LocalSenderMonitor`
- Add `MonitorMaxFloat()` as a method for `LocalSenderMonitor` and do not include it in the `SenderMonitor` interface
- Use `LocalSenderMonitor` in `Redeemer`

I see the following benefits from these changes:

- Avoids adding an additional method to the `SenderMonitor` API when it is not needed by any of the API consumers
- `redeemerClient` in #1513 does not need to implement `MonitorMaxFloat()`

If we use the approach described above, we end up with the following:
- The `SenderMonitor` interface is implemented by `LocalSenderMonitor` and `redeemerClient` (in #1513)
- `Redeemer` (in #1513) uses `LocalSenderMonitor` for ticket queue and max float monitoring functionality
- `LocalSenderMonitor` exposes max float monitoring functionality via a `MonitorMaxFloat()` (or `SubscribeMaxFloat()` if we’re using that naming convention) which is relevant for any user that needs access to max float updates as they happen (i.e. `Redeemer`), but that is *not* relevant for a user that uses the `SenderMonitor` interface
- `redeemerClient` receives max float values from the remote `Redeemer` in the background over the wire and returns the latest max float for a broadcaster via the `SenderMonitor` API method `MaxFloat()`
- Max float monitoring is abstracted away from the `SenderMonitor` API consumer since it is a detail of the `SenderMonitor` implementation. All the `SenderMonitor` API consumer is aware of is a `MaxFloat()` method that will return the latest max float for a broadcaster.

Reply:
sounds good to me, I believe 20ba922e covers the necessary changes in scope for this PR.

Reply:
Should we make `LocalSenderMonitor` implement `SenderMonitor` implicitly as well (return struct, not the interface from the constructor) ?

Reply:
Sure - don't see any reason why the interface should be returned here instead of the struct.